File: tree_iterator.py
# Author: Naveen US
# Title: Binary Search Tree Iterator

# Time complexity: O(h) 
# Space complexity: O(h)
# Did the code run on leetcode: Yes

class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
            self.stack = []
            while root:
                self.stack.append(root)
                root = root.left

    def next(self) -> int:
        node = self.stack.pop()
        if node.right:
            current = node.right
            while current:
                self.stack.append(current)
                current = current.left
        return node.val
        

    def hasNext(self) -> bool:
        return len(self.stack) > 0
        

File: tree-iterator.java
//Time Complexity: O(1)
//Space Complexity:o(1)
//Did this code successfully run on Leetcode :Yes

class BSTIterator {
    Stack<TreeNode> st;

    public BSTIterator(TreeNode root) {
        st = new Stack<>();
        dfs(root);
    }

    public int next() {
        TreeNode resultNode = st.pop();
        dfs(resultNode.right);
        return resultNode.val;
    }

    public boolean hasNext() {
        return !st.isEmpty();

    }

    public void dfs(TreeNode root) {
        while (root != null) {
            st.push(root);
            root = root.left;
        }
        return;
    }
}